File: BOJ 랜덤디펜스/5639.py
import sys
sys.setrecursionlimit(10000)
input = sys.stdin.readline

nums = []
while True:
    try:
        nums.append(int(input()))
    except:
        break

def postorder(s, e):
    if s > e:
        return
    mid = e + 1
    for i in range(s + 1, e + 1):
        if nums[i] > nums[s]:
            mid = i
            break
    postorder(s + 1, mid - 1)
    postorder(mid, e)
    print(nums[s])

postorder(0, len(nums) - 1)

File: BOJ/랜덤디펜스/5639.js
const input = require('fs').readFileSync('/dev/stdin', 'utf8').trim().split('\n').map(Number);
const postorder = (s, e) => {
  if (s > e) return;
  let mid = e + 1;
  for (let i = s + 1; i <= e; i++) {
    if (input[i] > input[s]) {
      mid = i;
      break;
    }
  }
  postorder(s + 1, mid - 1);
  postorder(mid, e);
  ans += input[s] + '\n';
};

let ans = '';
postorder(0, input.length - 1);
console.log(ans);
